File: harness/semgrep/fixtures/c1-gated/verify.good.ts
// GOOD: presence is asserted FIRST (throw when the field is missing), then the
// on-chain proof runs unconditionally. Omitting the field fails CLOSED.

interface Credential {
    signature?: string;
    payload: unknown;
}

async function verifyCredential(cred: Credential): Promise<boolean> {
    if (!cred.signature) {
        throw new Error('signature is required');
    }
    await verifyOnChain(cred.signature, cred.payload);
    return true;
}

async function verifyBare(signature: string | undefined, payload: unknown) {
    if (!signature) {
        throw new Error('missing signature');
    }
    await checkOnChainProof(signature, payload);
    return true;
}

File: harness/semgrep/fixtures/c1-gated/verify.good.py
# GOOD: assert presence first, then verify unconditionally.


def verify_credential(cred):
    if not cred.signature:
        raise ValueError("signature is required")
    verify_on_chain(cred.signature, cred.payload)
    return True


def verify_bare(signature, payload):
    if not signature:
        raise ValueError("missing signature")
    check_on_chain_proof(signature, payload)
    return True
